rfd/0030-upgrade-system.md

```diff
@@ -0,0 +1,444 @@
+with one final scenario: the index *and* at least one repository are compromised simultaneously (but
+the compromise is not so severe as to put us in the previously discussed 'doomsday scenario').  This is
+a bit niche, but lets exercise a little professional paranoia.  We should have a mechanism of validating
+package version (prior to unpacking/installation) as part of package signature verification (i.e. on the
+"distribution" side).  One appealing option is [minisign](https://jedisct1.github.io/minisign/) which
+is a simple and modern signing utility that supports "trusted comments" (i.e. signed metadata) as part
+of the signature file.
+
+
+#### Compromised Build System
+
+Out of scope for now, but I've heard all the cool kids are doing deterministic builds these days.
+
+
+## Proposal
+
+### Overview
+
+High-level overview of system components, including discussion of MVP vs planned features.
+
+#### Index Server
+
+The index server, as discussed above, is responsible for serving metadata and checksums for the
+current list of valid installation targets.  Part of the security model of the upgrade system is
+that the index server remains isolated from the rest of the ugrade system stack, and provides only
+abstract information about available releases (i.e. it will not provide things like download urls
+or install scripts).
+
+Key design considerations related to the index server are as follows:
+- Index server should be trivially horizontally scalable.
+- Index servers should always require TLS 1.3 unless started in development mode.
+- Rollbacks of index server state or API version must not be detremental to its clients
+(i.e. clients must be resilient to inconsistent values being returned).
+- Index server must remain compatible with many past teleport versions (exact number TBD),
+but we don't want to orphan older versions, so lets err on the side of caution here.
+
+Outstanding design questions:
+- Grpc or basic https API?  My preference leans toward grpc, but its hard to discount the possible
+benefits of a basic https API (e.g. pulling down release info with `curl`).
+- How 'smart' should the index server be exactly?  On one hand, if the index server is able to intelligently
+select what releases/targets are relevant to a given client, it may be able to reduce the impact of bugs
+in the upgrade system (easier to update the index server than to update all clients).  On the other hand,
+the extensibility of the upgrade system (not to mention the simplicity of the index server) would likely
+be benifited by keeping as much logic as possible in the client clusters.
+
+Hypothetical teleport release descriptors returned by index server:
+
+```yaml
+targets:
+  - tags: [oss]
+    version: '100.1.0-alpha.2'
+    arch: amd64
+    os: linux
+    flavor: oss
+    stable: false
+    security_patch: false
+    deprecated: false
+    sums:
+      rpm-blake2-256: '...'
```

Comment:
What kind of signatures are we going to support initially and why?

Reply:
Release manifests like the one shown here provide hashes, not signatures (writing a manifest shouldn't require a secret).

That being said, I'm sold on your suggestion of using `minisign` for package signing.  Added a crypto section in most recent commit.